### PR TITLE
[INLONG-1992] sort-flink support configurable loader of getting configuration.

### LINKED ISSUE #1992: [Feature]sort-flink support configurable loader of getting configuration.
### Description

sort-flink support configurable loader of getting configuration.

### Use case

_No response_

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by luchunliang**
### Title Name: [INLONG-1992][Inlong-sort-flink] sort-flink support configurable loader of getting configuration.

Fixes #1992 

### Motivation

*Explain here the context, and why you're making that change. What is the problem you're trying to solve.*

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Review by gosonzhang — APPROVED** [review]
+1

**Comment by dockerzhang** [comment]
@ifndef-SleePy PTAL

**Review by dockerzhang — APPROVED** [review]

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/ZookeeperWatcher.java [hunk 1]
```diff
@@ -38,10 +39,16 @@
...
+import org.apache.inlong.sort.configuration.Constants;
+import org.apache.inlong.sort.meta.MetaManager.DataFlowInfoListener;
+import org.apache.inlong.sort.util.ZooKeeperUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public class ZookeeperWatcher implements AutoCloseable, UnhandledErrorListener {
+import com.google.common.base.Preconditions;
+
+public class ZookeeperWatcher implements AutoCloseable, UnhandledErrorListener, MetaWatcher {
```
Comment (chantccc) [inline inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/ZookeeperWatcher.java]: ZookeeperWatcher is a basic util class wrapping zk related operations. It should not implement MetaWatcher.
Use another class ZookeeperMetaWatcher which implements MetaWatcher instead.

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/ZookeeperWatcher.java [hunk 2]
```diff
@@ -50,26 +57,52 @@
...
-    private final Map<String, PathChildrenCache> pathChildrenCaches;
+    private Map<String, PathChildrenCache> pathChildrenCaches;
 
-    private final ConnectionStateListener connectionStateListener =
-            (client, newState) -> handleStateChange(newState);
+    private ConnectionStateListener connectionStateListener = (client, newState) -> handleStateChange(newState);
 
-    public ZookeeperWatcher(CuratorFramework client) {
-        this.client = checkNotNull(client, "CuratorFramework client");
+    /**
```
Comment (chantccc) [inline inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/ZookeeperWatcher.java]: descriptions of param are empty.

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/MetaWatcher.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+/**
+ * 
+ * MetaWatcher
+ */
+public interface MetaWatcher {
+
+    /**
+     * open
+     * 
+     * @param  config
```
Comment (chantccc) [inline inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/MetaWatcher.java:32]: descriptions of param are empty.

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/DataFlowsChildrenWatcherListener.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+import org.apache.inlong.sort.protocol.DataFlowInfo;
+import org.apache.inlong.sort.protocol.DataFlowStorageInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * 
+ * DataFlowsChildrenWatcherListener
+ */
+public class DataFlowsChildrenWatcherListener implements ChildrenWatcherListener {
```
Comment (chantccc) [inline inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/DataFlowsChildrenWatcherListener.java]: This class is only used in ZookeeperMetaWatcher, make it an inner class is better.

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/ZookeeperWatcher.java (hunk 1 above)
Comment (luchunliang) [reply inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/ZookeeperWatcher.java]: ok, add another class ZookeeperMetaWatcher  which implements MetaWatcher, and change ZookeeperWatcher  to ZookeeperWatcherUtils.

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/ZookeeperWatcher.java (hunk 2 above)
Comment (luchunliang) [reply inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/ZookeeperWatcher.java]: add param comment.

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/MetaWatcher.java (hunk 3 above)
Comment (luchunliang) [reply inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/MetaWatcher.java:32]: add param comment.

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/DataFlowsChildrenWatcherListener.java (hunk 4 above)
Comment (luchunliang) [reply inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/DataFlowsChildrenWatcherListener.java]: ok, make it to an inner class

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/zookeeper/ZookeeperWatcherUtils.java [hunk 5]
```diff
@@ -38,38 +39,74 @@
...
+     * @param  metaListener a listener of DataFlowInfo
+     * @throws Exception    any exception
+     */
+    public void open(Configuration config, DataFlowInfoListener metaListener) throws Exception {
+        this.client = ZooKeeperUtils.startCuratorFramework(config);
         this.caches = new HashMap<>();
         this.pathChildrenCaches = new HashMap<>();
         client.getUnhandledErrorListenable().addListener(this);
         client.getConnectionStateListenable().addListener(connectionStateListener);
+        synchronized (MetaManager.LOCK) {
```
Comment (chantccc) [inline inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/zookeeper/ZookeeperWatcherUtils.java:90]: Move these codes into ZookeeperMetaWatcher.
ZookeeperWatcherUtils should not be aware of anything about dataflow.

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/zookeeper/ZookeeperWatcherUtils.java [hunk 6]
```diff
@@ -38,38 +39,74 @@
...
+                dataFlowsChildrenWatcherListener.onInitialized(childData);
+            }
+        }
+    }
+
+    /**
+     * If you want to change the path rule here, please change ZkTools#getNodePathOfDataFlowStorageInfoInCluster in api
+     * too.
+     */
+    public static String getWatchingPathOfDataFlowsInCluster(String cluster) {
```
Comment (chantccc) [inline inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/zookeeper/ZookeeperWatcherUtils.java:108]: ditto

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/zookeeper/ZookeeperWatcherUtils.java [hunk 7]
```diff
@@ -337,4 +377,95 @@ public void childEvent(
             }
         }
     }
+
+    private static class DataFlowsChildrenWatcherListener implements ChildrenWatcherListener {
```
Comment (chantccc) [inline inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/zookeeper/ZookeeperWatcherUtils.java:381]: ditto

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/MetaManager.java [hunk 8]
```diff
@@ -132,149 +129,116 @@ public void registerDataFlowInfoListener(DataFlowInfoListener dataFlowInfoListen
...
-                    LOG.info("Try to update dataFlow {}.", dataFlowId);
-
-                    for (DataFlowInfoListener dataFlowInfoListener : dataFlowInfoListeners) {
-                        try {
-                            dataFlowInfoListener.updateDataFlow(dataFlowInfo);
-                        } catch (Exception e) {
-                            LOG.warn("Error happens when notifying listener data flow updated", e);
-                        }
-                    }
+        DataFlowInfo oldDataFlowInfo = dataFlowInfoMap.put(dataFlowId, dataFlowInfo);
```
Comment (chantccc) [inline inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/MetaManager.java:184]: ditto

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/MetaManager.java [hunk 9]
```diff
@@ -132,149 +129,116 @@ public void registerDataFlowInfoListener(DataFlowInfoListener dataFlowInfoListen
...
+    /**
+     * removeDataFlow
+     * 
+     * @param  dataFlowInfo
+     * @throws Exception
+     */
+    public void removeDataFlow(DataFlowInfo dataFlowInfo) throws Exception {
+        long dataFlowId = dataFlowInfo.getId();
+
+        dataFlowInfoMap.remove(dataFlowId);
```
Comment (chantccc) [inline inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/MetaManager.java:222]: ditto

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/zookeeper/ZookeeperWatcherUtils.java (hunk 5 above)
Comment (luchunliang) [reply inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/zookeeper/ZookeeperWatcherUtils.java:90]: ZookeeperMetaWatcher implements MetaWatcher.
ZookeeperWatcherUtils is a basic util class.

chantccc yesterday
ZookeeperWatcher is a basic util class wrapping zk related operations. It should not implement MetaWatcher.
Use another class ZookeeperMetaWatcher which implements MetaWatcher instead.

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/zookeeper/ZookeeperWatcherUtils.java (hunk 6 above)
Comment (luchunliang) [reply inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/zookeeper/ZookeeperWatcherUtils.java:108]: ZookeeperMetaWatcher implements MetaWatcher.
ZookeeperWatcherUtils is a basic util class.

chantccc yesterday
ZookeeperWatcher is a basic util class wrapping zk related operations. It should not implement MetaWatcher.
Use another class ZookeeperMetaWatcher which implements MetaWatcher instead.

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/zookeeper/ZookeeperWatcherUtils.java (hunk 7 above)
Comment (luchunliang) [reply inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/zookeeper/ZookeeperWatcherUtils.java:381]: ZookeeperMetaWatcher implements MetaWatcher.
ZookeeperWatcherUtils is a basic util class.

chantccc yesterday
ZookeeperWatcher is a basic util class wrapping zk related operations. It should not implement MetaWatcher.
Use another class ZookeeperMetaWatcher which implements MetaWatcher instead.

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/MetaManager.java [hunk 10]
```diff
@@ -132,149 +129,116 @@ public void registerDataFlowInfoListener(DataFlowInfoListener dataFlowInfoListen
...
-                return;
+    /**
+     * addDataFlow
+     * 
+     * @param  dataFlowInfo
+     * @throws Exception
+     */
+    public void addDataFlow(DataFlowInfo dataFlowInfo) throws Exception {
+        long dataFlowId = dataFlowInfo.getId();
+        DataFlowInfo oldDataFlowInfo = dataFlowInfoMap.put(dataFlowId, dataFlowInfo);
```
[the remark this answers is not in the document]
Comment (luchunliang) [reply inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/MetaManager.java:152]: add LOCK operation.

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/MetaManager.java (hunk 8 above)
Comment (luchunliang) [reply inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/MetaManager.java:184]: add LOCK operation.

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/MetaManager.java (hunk 9 above)
Comment (luchunliang) [reply inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/MetaManager.java:222]: add LOCK operation.

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/zookeeper/ZookeeperWatcherUtils.java (hunk 5 above)
Comment (chantccc) [reply inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/zookeeper/ZookeeperWatcherUtils.java:90]: Basic util class means  that ZookeeperWatcherUtils is just a wrapper of curator, all business codes should not be placed in it.
The watch of dataflow should be placed in ZookeeperMetaWatcher(in our case,  data of dataflow is called meta)

Comment (luchunliang) [reply inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/zookeeper/ZookeeperWatcherUtils.java:90]: ok, move those codes of about DataFlowInfo to ZookeeperMetaWatcher.

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/zookeeper/ZookeeperWatcherUtils.java (hunk 6 above)
Comment (luchunliang) [reply inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/zookeeper/ZookeeperWatcherUtils.java:103]: ok, move those codes of about DataFlowInfo to ZookeeperMetaWatcher.

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/zookeeper/ZookeeperWatcherUtils.java (hunk 7 above)
Comment (luchunliang) [reply inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/meta/zookeeper/ZookeeperWatcherUtils.java]: ok, move those codes of about DataFlowInfo to ZookeeperMetaWatcher.

**Comment by chantccc** [comment]
+1

**Pull request merged**